### ISSUE TITLE
[Feature propose] versioning in pact, and pact upgrade & pact dist-upgrade (or the other way)

### ISSUE THREAD

**Issue by v-yadli** [babun/babun#239]
We'll have to come up with something in order to stop the confusion about commands yielding 127 silently.

The root case is that now dependency is handled without considering the versions of packages. However, Cygwin packages update with a dependency of constantly updating low level libraries such as libgcc1 (this package has been mentioned in quite a few issues) and unfortunately these low level libraries have no compatibility across different versions at all.

luckily we could reconstruct the dependency tree with "compatible version spans" (think about it like a dep tree with a slider on each tree node, allowing some interchangeable versions to present), and all the packages captured in a `pact update` command are assumed to be compatible with each other.

Thus, if the low level packages are not changes across `pact update` repo snapshots, we could say that the packages are compatible with the previous versions, and thus we could code up 'pact upgrade'.

Also, if the low level packages are changed, we could say that it's a "dist-upgrade" like going from Fancy_Distro_Name X to  Fancy_Distro_Name X+1. An external program (not relying on the Cygwin env to be upgraded) is needed though.

This will stop the problem on the Cygwin side, but not those real package dependency problems.
However imho Cygwin is not designed to be used like this anyway. We might work it out the other way by not fetching a new package list that frequently. It's like to drop 'pact upgrade' completely and leaves only `pact dist-upgrade`.

==========================

update: related issues:

#235 #232 #231 #229 #225 #222 #210 #203 (and possibly more)

**Comment by tombujok** [comment]
Thank you for this extremely constructive voice in the discussion about babun updates!
When I released babun last year I knew that this problem would eventually pop-up, but I had deferred it till the moment when it actually becomes a pain in the neck... As you said, it's the perfect moment to solve it now.

Would it be possible to schedule a skype call where we could discuss the possible solutions and decide to implement one of them? All people who are interested in this topic are welcome to participate. 
@v-yadli @jlupi What do you think?

**Comment by v-yadli** [comment]
Sure! :) Where are you guys? I'm in CST +8. It's Chinese new year here and it's no problem to arrange it in the working hours.

Sent from my Sony Xperia™ smartphone

---- Tom Bujok wrote ----

Thank you for this extremely constructive voice in the discussion about babun updates!
When I released babun last year I knew that this problem would eventually pop-up, but I had deferred it till the moment when it actually becomes a pain in the neck... As you said, it's the perfect moment to solve it now.

Would it be possible to schedule a skype call where we could discuss the possible solutions and decide to implement one of them? All people who are interested in this topic are welcome to participate.
@v-yadli @jlupi What do you think?



---

**Comment by v-yadli** [comment]
@tombujok after thinking about this again I feel that leaving dist-upgrade only would be a better idea. Less effort and more consistency. Please drop me a mail if you'd like to schedule a meeting.

**Comment by tombujok** [comment]
Thanks! I was also thinkg about it  and the `dist-upgrade` options seems the only reasonable solution.
`dist-updrade` could be implemented in the following way:
- in babun we keep track of the underlying cygwin version (e.g. using their versioning 1.7.35)
- we also keep track of the newest cygwin version, e.g. like here, but in the cygwin.version file:
https://github.com/babun/babun/blob/master/babun.version
- babun update compares the current and the newest cygwin version - in the same way as it already does with the babun's version.
- whenver a new cygwin version is released we increment the marker in the cygwin.version file, so that the babun check command may printa message that a dist-updrade is necessary
- whenever a pact install / uninstall command is executed the cygwin's versions are also compared, not allowing the user to install a new package if he/she is not running on the newest cygwin version
- we also code up a upgrade.bat file that could be executed to execute the dist-upgrade 

The upgrade.bat file would:
- download the newest  setup-x86.exe from cygwin.com
- execute the cygwin update silently pointing to the babun's cygwin folder

Open questions:
1.) Does the dist-upgrade looking like above make sense to everybody? Am I missing something here?
2.) Would the setup-x86.exe dist-upgrade silently recognize and upgrade all packages that have been already installed with pact?
3.) Do also we show a popup message on babun.bat start asking whether the user would like to do the dist-upgrade?

**Comment by v-yadli** [comment]
"•whenever a pact install / uninstall command is executed the cygwin's versions are also compared, not allowing the user to install a new package if he/she is not running on the newest cygwin version"

This is a solid way to guarantee consistency. However, it might occur too frequently that it would break the whole work flow. How about still letting a user to install from the old repo package list file? The old packages are very likely to be still there (and that's how I recovered from mistakenly installed new packages at first). It is indeed necessary to inform users that Cygwin has moved forward so the current snapshot would be dropped indefinitely.

I'm playing around the installer and see if it does the job for us...

**Comment by v-yadli** [comment]
"2.) Would the setup-x86.exe dist-upgrade silently recognize and upgrade all packages that have been already installed with pact?"

Yes please see -g option.
The installer is quite nice :-)

**Comment by tombujok** [comment]
ad.1) But how would we know that the older versions of installed packaged are compatible with the older cygwin dist?
ad.2) Would you paste a "command" that does the upgrade silently assuming that the setup-x86.exe is in the current folder?

**Comment by v-yadli** [comment]
@tombujok Could you please drop me a mail to yatli@microsoft.com? There's a very interesting solution I'd like to discuss with you. The basic idea is that we can setup a server to record all the changes from Cygwin. And babun users can then use RESTful API to query the server and pull back dependency information.

**Comment by tombujok** [comment]
@v-yadli Email sent! 
What I am afraid of is that we will intensively rely on our own logic and that we will make assumptions about Cygwin that will not always hold - so that the whole upgrade system may collapse at the end...
For example when somebody upgrades from version X to version X+3 (skipping 2 version inbetween).

Cygwin guys frequently change things. If we could rely on their setup.exe upgrade mechanism we would be 100% sure that everything works in every update... WDYT?

**Comment by jlupi** [comment]
@tombujok this dist-upgrade looks pretty good to me 

> whenever a pact install / uninstall command is executed the cygwin's versions are also compared, not allowing the user to install a new package if he/she is not running on the newest cygwin version

>> This is a solid way to guarantee consistency. However, it might occur too frequently that it would break the whole work flow. How about still letting a user to install from the old repo package list file? 

>>> But how would we know that the older versions of installed packaged are compatible with the older cygwin dist?

Both are very good points. We should consider what is the better option taking into account that there were about 8-10 cygwin releases in the past year
1.7.31 Jul 2014 https://cygwin.com/ml/cygwin-announce/2014-07/msg00010.html
1.7.32 Aug 2014 https://cygwin.com/ml/cygwin-announce/2014-08/msg00020.html
1.7.33 Oct 2014 https://cygwin.com/ml/cygwin-announce/2014-10/msg00042.html
1.7.34 Nov 2014 https://cygwin.com/ml/cygwin-announce/2014-11/msg00020.html
1.7.35 Mar 2015 https://cygwin.com/ml/cygwin-announce/2015-03/msg00009.html

It would be interesting to know if Red-Hat has a fixed schedule for cygwin releases or if it rather happens ad-hoc. 

> What I am afraid of is that we will intensively rely on our own logic and that we will make assumptions about Cygwin that will not always hold

I would also rather rely on cygwin's installer for now (also to avoid complexity). 
However there are some requests to support a downgrade of packages. Looks like with this solution we'll rather not be able to do that (it might be necessary to downgrade the whole dist).
@v-yadli would your solution be able to support this?

**Comment by v-yadli** [comment]
@jlupi Nak I discussed with @tombujok and I realized that it will not work as expected given that old Cygwin packages will be removed from the repo.

However, as per:
http://cygwin.com/faq/faq.html#faq.setup.cli


>2.6.
> 
>Can I use Cygwin Setup to get old versions of packages (like gcc-2.95)?
>
> 
>Cygwin Setup can be used to install any packages that are on a Cygwin mirror, which usually >includes one version previous to the current one. The complete list may be searched at https://cygwin.com/packages/. There is no complete archive of older packages. If you have a problem with the current version of a Cygwin package, please report it to the mailing list using the guidelines at https://cygwin.com/problems.html. 
>
>That said, if you really need an older package, you may be able to find an outdated or archival mirror by searching the web for an old package version (for example, gcc2-2.95.3-10-src.tar.bz2), but keep in mind that this older version will not be supported by the mailing list and that installing the older version will not help improve Cygwin. 
 
If I understand correctly, even if the old packages are not binary compatible with the new ones, one can still compile them from source. Yet the problem is still that older packages are to be "searched over the internet"...

**Comment by v-yadli** [comment]
The installer automatically updates with the following command:

`setup-x86.exe -s http://mirrors.kernel.org/sourceware/cygwin  -R "D:\your_cygwin_root" -g -N -d -q -a x86 -B`

However the GUI is still shown and even with "quiet mode" (`-q`) it still won't go away.
Is there any general way of disabling a window?

**Comment by tombujok** [comment]
Thanks! We hacked on it a bit at Hackergarten on Tuesday and @airbornpl pushed a pull request that:
https://github.com/babun/babun/pull/246

It contains the script that will do the dist-upgrade. @v-yadli Could you review it? 
Unfortunately it's impossible to get rid of the window completely, but the user will not be asked anything, and Admin rights are not required, so I think it's OK...

**Comment by v-yadli** [comment]
Awesome! I just added a quick note on the .bat file.
As for the GUI I agree with you. It does provide some nice progress indicators :-)

**Comment by tombujok** [comment]
OK. The fix is ready. The next release is scheduled for around Wednesdey (if tests are successful).
I need to add a Yes/No popup to the update.bat script. I was thinking about sth like this:
http://stackoverflow.com/questions/23530293/adding-vbs-yes-no-box-code-to-a-batch-file

Does anybody know if it works on all Windows versions? Or is there a better way to do it?

**Comment by v-yadli** [comment]
http://www.autohotkey.com/docs/commands/MsgBox.htm

Guess this will be more robust and scalable, and it can be compiled into a portable executable.

**Comment by duck** [comment]
I have a few errors I am seeing.

First, while the wget uses the Proxy as defined, the actual setup-x86.exe does not, and I have no way of passing proxy info to it w/o hacking the .bat file.

Also, I am seeing this error:
```
Starting cygwin install, version 2.870
User has NO backup/restore rights
io_stream_cygfile: fopen(/etc/setup/setup.rc) failed 2 No such file or directory
Current Directory: C:\Users\harperdo\.babun\dist
```
This is running as a domain user under Windows 8.1

**Comment by duck** [comment]
adding ```--proxy=myproxyhost:8080``` to the script does the trick.

**Comment by tombujok** [comment]
@duck  Which script are you running?

**Comment by duck** [comment]
update.bat from pull request #246.

**Comment by tombujok** [comment]
Ok, thankx. @jlupi still is working on the script. We have experienced some other problems to...

**Comment by jlupi** [comment]
I recently updated the update.bat file.

The bat script now additionally checks whether the user has http_proxy set in ~./babunrc and passes it to the cygwin installer (only host:port, the user has to provide his username and password manually)

I also removed the --download option. The download option silently sets 'download without installing' option on the installer but what we really want is the 'Install from Internet' option.

![image](https://cloud.githubusercontent.com/assets/4189968/6777514/a40706f6-d14a-11e4-8b11-81d6acd61afd.png)

The problem here is that the installer stores those options in the windows registry. This means that if you once selected the 'download without installing' option and re-run the installer without --download in silent mode it will still remember this setting. I don't know if there is any way to force the 'download from Internet' setting. At least I do not see any flag you can set in the installer.

If you have previously selected the 'Install from Internet' and re-run the update.bat in quiet mode than it works as expected (I tried it with and without proxy).

For me at the moment the possible solutions are:
- set a flag to cygwin installer to enforce 'install from internet' option -> prefered but probably not possible 
- disable the --quiet-mode option and let user clicking through the installation process -> no hacks but quite painful for the user
- reset cygwin settings in windows registry (o_O)
- any other?

**Comment by tombujok** [comment]
I found sth like this here: http://superuser.com/questions/167219/how-to-completely-uninstall-remove-cygwin-from-windows-7

*"Finally, if you want to be thorough you can delete the registry tree Software\Cygwin under HKEY_LOCAL_MACHINE and/or HKEY_CURRENT_USER."*

Can anybody script it?

**Comment by v-yadli** [comment]
I guess the entries should not be deleted, or otherwise the installer will not detect the existing installation.
However, I checked my registry and there're no entries related to download options.

The entry is like:
```
key = 1299052366339799
value = \??\D:\yatli\.babun\cygwin
```

Options encoded in the key?

**Comment by tombujok** [comment]
The installer will recognize the existing installation as we pass all the info in the cmd-line.

The entries are stale -> they are there only if you have a separate cygwin instance.
So, we don't need them for babun - but if somebody has an existing cygwin instance we will screw it, if we delete this entries...

Probably the only solution is to pass all the options to the update.bat / cygwin.exe, but the user will have to click "OK" 3 times during the dist-upgrade...

**Comment by v-yadli** [comment]
Would you mind using an ahk script? It clicks on "OK" faster than one can perceive.

**Comment by tombujok** [comment]
Awesome! Could you create a proof-of-concept script to test the update.bat?

**Comment by v-yadli** [comment]
alright! this is fun.. :-D

**Comment by v-yadli** [comment]
I think I'm missing something... This is the command line I'm using:
`setup-x86.exe --upgrade-also --site="http://mirrors.kernel.org/sourceware/cygwin" --no-admin --no-shortcuts --no-startmenu --no-desktop --root="D:\TestCygwin" --local-package-dir="D:\temp\packages"`

However it completes the installation silently without any message boxes popping out.
Manually ran setup-x86 once and set it to download only, and then the command line above also switched to download only.

removing --quiet-mode will bring more than just a few "OK"s to click through... So what is the correct way to bring up the message boxes to click?

**Comment by tombujok** [comment]
hmm, so without `--quiet-mode` you still do not get the wizard steps to confirm?
@jlupi could you help?

**Comment by v-yadli** [comment]
I mean, I can get the wizard, but there are a lot of options to go through just like a standard installation. In that case the robustness of ahk cannot be guaranteed.

**Comment by tombujok** [comment]
OK, but the thing is that with all the options that you specify in the command line, you see 3 or 4 wizard steps, but all the options are already preselected (mirror, all packages, upgrade, etc.) You just have to make sure that the "Install from the Internet" is selected.", plus "Use Internet Explorer Proxy".
Is that possible?

**Comment by v-yadli** [comment]
hmm. Here's what I have in my mind now:

Method 1:

the first radiobox page "Install from the Internet" can be secured with several up arrow keys then enter

the second  "Use Internet Explorer Proxy" alt+i then enter

then wait for the window to resize...

then wait for the installation to finish (no reliable way to do so!)

Method 2:

run Setup-x86.exe, select  "Install from the Internet",  "Use Internet Explorer Proxy", then exit right away (enter up up up enter enter enter alt+i esc enter), the options will be recorded. (Amazing!)

Then, start up the unattended version... XD

I'm going to try method 2 right away..

**Comment by v-yadli** [comment]
It works...

The bat file content:

```
start install.ahk
setup-x86.exe --upgrade-also --site="http://mirrors.kernel.org/sourceware/cygwin" --no-admin --no-shortcuts --no-startmenu --no-desktop --root="D:\TestCygwin" --local-package-dir="D:\temp\packages"
setup-x86.exe --quiet-mode --upgrade-also --site="http://mirrors.kernel.org/sourceware/cygwin" --no-admin --no-shortcuts --no-startmenu --no-desktop --root="D:\TestCygwin" --local-package-dir="D:\temp\packages"
```

install.ahk content:
```ahk
WinWaitActive , Cygwin Setup, , 100
Send {Enter}
Sleep 1
Send {Up}
Sleep, 1
Send {Up}
Sleep, 1
Send {Up}
Sleep, 1
Send {Enter}
Sleep, 1
Send {Enter}
Sleep, 1
Send {Enter}
Sleep, 1
Send !i
Sleep, 1
Send {Esc}
Sleep, 1
Send {Enter}
```

**Comment by tombujok** [comment]
Awesome!! Could you send a pull request with the adjusted update.bat script and the ahk.exe (in the tools folder), so that everything is bundled and can be automated from within's update.bat

**Comment by v-yadli** [comment]
see #254 
I didn't test it because update.bat complains about not finding wget.exe. Could you check if it works?

**Comment by tombujok** [comment]
Thx! I've merged it and I will definitely test too.
It should also work for you if you take the existing .babun folder and copy the akh.exe to tools and the update.bat file to the .babun folder (where the babun.bat file resides)

**Comment by v-yadli** [comment]
I've put the tools to the correct place and tested it working. The only problem now is that the send key mechanism is not robust. If any window jumps out and grabs the focus this little script will fail. Let me think if I can get a workaround..

**Comment by tombujok** [comment]
OK, it would be great to make it robust, so that we may bake the newest release ;) Thanks to your awesome work we are close to it!

**Comment by v-yadli** [comment]
An updated script:

```
WinWaitActive , Cygwin Setup, , 100
WinActivate ,Cygwin Setup
SetKeyDelay, 0, 0, Play
ControlSend ,Button2,{Enter}, Cygwin Setup
ControlSend ,,{Up}, Cygwin Setup
ControlSend ,,{Up}, Cygwin Setup
ControlSend ,,{Up}, Cygwin Setup
ControlSend ,Button7,{Enter}, Cygwin Setup
ControlSend ,Button12,{Enter}, Cygwin Setup
ControlSend ,Button14,{Enter}, Cygwin Setup
Sleep, 1
ControlSend ,Button2,{Alt down}i{Alt up}, Cygwin Setup
ControlSend ,Button2,{Esc down}{Esc up}, Cygwin Setup
WinWaitActive , Exit Cygwin Setup?, , 1
WinActivate ,Exit Cygwin Setup?
ControlSend ,,{Enter}, Exit Cygwin Setup?
```

This one uses ControlSend to precisely send into the buttons.
The `Sleep, 1` is critical or there's 20% chance to choke the installer.
I'm issuing a new pull request.

**Comment by v-yadli** [comment]
please see #255. This one works much better. :-)

**Comment by jlupi** [comment]
@tombujok 

> So, we don't need them for babun - but if somebody has an existing cygwin instance we will screw it, if we delete this entries...

We would rather not screw up someone's cygwin installation if we delete those keys. 
As you can read in https://cygwin.com/faq/faq.html#faq.setup.cli 

> How do I save, restore, delete, or modify the Cygwin information stored in the registry?
>> Since Cygwin 1.7, there's nothing important in the registry anymore, except for the installation directory information stored there for the sake of setup-x86{_64}.exe. There's nothing left to manipulate anymore.

Looks like in worst case when you start the cigwin installer in gui mode it will not remember your previous settings.

I didn't try @v-yadli 's script yet though, will try it soon.

**Comment by tombujok** [comment]
OK. Please test it - you can also try with deleting the registry entries... Maybe it would work?

**Comment by jlupi** [comment]
I tried deleting registry keys but it didnt help. The state must be persisted somewhere on the filesystem. I'll try to check on the installer's source code.

**Comment by tombujok** [comment]
OK guys, the situation looks like this:
- we still cannot find the place where the setup.exe installer stores the "install-switch" if to "Download from the Internet" or "Download without installing". We know for sure that it's not stored in the windows registry. We hjave deleted all the registry entries, started setup.exe, chose one option, clicked next and it was stored, although there was nothing in the registry...
- akh solution is OK, but the best option would be to avoid it - it's still the second choice option, since there are different paths, for example with proxy, and then the script does not work. Thanks to @v-yadli for his efforts though. We still may use it!

The preferred solution is:
- to find the place where the "Install-switch" is stored 
- to delete that "install-switch" in the update.bat script
- and to run the installer in the quite-mode

In order to achieve that we started browsing the cygwin's setup.exe source code (sic!)
Any help is appreciated!
git clone git://cygwin.com/git/cygwin-setup.git

**Comment by v-yadli** [comment]
@tombujok Did you try capture all the IO activities of the installer with procmon? maybe that will help ;)

**Comment by tombujok** [comment]
No, but it would be great if somebody could try it out! 
@v-yadli There's already evening in your timezone, we have just started working ;)

The path to check is:
- start setup.exe
- choose "Download from the Internet"
- click "Next"
- close the installer
That's enough for setup.exe to persist that "install-switch".

**Comment by jlupi** [comment]
I found it, the UserSettings.cc loads the settings file setup.rc from cygwin's from /etc/setup folder

**Comment by jlupi** [comment]
I executed the installer several times with diferent options and this is how the setup.rc file is modified:

The 'Install from Internet' option writes

````
last-action
	Download,Install
````

where the 'download without installing' writes

````
last-action
	Download
````

*Proxy Settings*

User Proxy

````
net-method
	Proxy
net-proxy-host
	webproxy.xxx.com
net-proxy-port
	8080
````

Use Internet Explorer Proxy settings

````
net-method
	IE
net-proxy-host
	webproxy.xxx.com
net-proxy-port
	8080
````

Direct

````
net-method
	Direct
net-proxy-host
	webproxy.xxx.com
net-proxy-port
	8080
````

The installer remembers the host and port but probably does not use it for 'Direct' or 'Internet Explorer' proxy

**Comment by jlupi** [comment]
looks like we can simplify our update script. It would be enough to create a setup.rc template and copy it to /etc/setup folder before the upgrade proces starts

**Comment by v-yadli** [comment]
@jlupi Ahh, yesterday I was messing with UserSettings.cc looking for registry operations and overlooked the /etc thing!

Maybe a cat in a bash script will fulfill the purpose of template generation?

**Comment by tombujok** [comment]
@jlupi You've unlocked the C-mastha badge, whereas @v-yadli unlocked the AKH-mastha badge ;)

OK, I would propose the following, we drop the akh solution;  in update.bat, we:
- delete the existing setup.rc file
- create a new setup.rc file with 
```
last-action
    Download,Install
net-method
    IE
```

I would also leave the existing /proxy switch in the update.bat.
When the user specifies it he/she will use the manually specified proxy (the password popup may be shown, but it's not a problem IMHO) WDYT?

**Comment by jlupi** [comment]
@tombujok sounds good. Would you rather leave the --site option as it is or shall we also put it into the setup.rc file?

@v-yadli what do you mean with cat in bash? I was thinking of a static file that will go to the tools folder and would just do a copy in the bat file. What do you think?

**Comment by v-yadli** [comment]
@jlupi No pun here (but cat in fish is more awesome ;), but if the proxy host name and port should be passed in, it would be better generated within a bash shell script that `cat` the static part and `echo` the dynamic part. WDYT?

**Comment by tombujok** [comment]
Yes, the update.bat should have /proxy (/p) and /site (/s) switches.
They will be propagated to setup.rc anyway, but we don't need to put them there, don't we?
@jlupi Could you finish it up and submit a pull request?
We are an "inch" away from a super cool candidate release... :+1:

**Comment by jlupi** [comment]
in the current script we do not have the /proxy switch. We need to add it.

The workflow would look like this:

if user provids proxy with /proxy switch we use this
else if we find proxy in babunrc we take it
else we set IE as a net-method in the setup.rc

I have quite a busy day today and I am gone in the evening so I can start working on it no sooner than tomorrow / thursday.

@v-yadli, @tombujok (or others) any chance you guys could start with it sooner?

@v-yadli no fish support yet :-) It is OK for me.

**Comment by v-yadli** [comment]
@jlupi Nak it's 11:00PM and I'm still at office. I'll see if I could quickly brush it up tmr morning.

**Comment by jlupi** [comment]
@v-yadli  easy, get some rest :)

I just tried the simplified setup.rc, it looks good I guess.

The only problem I have is that the installer keeps complaining about corrupted setup.ini file. I had to start it with the --no-verify option. Would it be save to use it in the script ?

I just discoved two problems (not related to the setup script itself). When i started the upgrade having babun running I got

````
The following DLLs couldn't be rebased because they were in use:
  /usr/lib/zsh/5.0.6/zsh/zutil.dll
  /usr/lib/zsh/5.0.6/zsh/zle.dll
  /usr/lib/zsh/5.0.6/zsh/terminfo.dll
  /usr/lib/zsh/5.0.6/zsh/stat.dll
  /usr/lib/zsh/5.0.6/zsh/parameter.dll
  /usr/lib/zsh/5.0.6/zsh/computil.dll
  /usr/lib/zsh/5.0.6/zsh/complist.dll
  /usr/lib/zsh/5.0.6/zsh/complete.dll
  /usr/bin/libzsh-5.0.6.dll
  /usr/bin/cygncursesw-10.dll
  /usr/bin/cyggcc_s-1.dll
````

Do we want to do anygthing about that?

Aftewwards I started it again but now i am getting 

````
Updating package information in /var/cache/rebase/rebase_pkg.
Updating rebase information for installed dynamic objects in /var/cache/rebase/rebase_lst.
/bin/rebaselst: 211: /bin/rebaselst: cannot create /var/cache/rebase/rebase_lst: Permission denied
Updating rebase information for installed executables in /var/cache/rebase/rebase_exe.
/bin/rebaselst: 197: /bin/rebaselst: cannot create /var/cache/rebase/rebase_exe: Permission denied
removing /var/cache/rebase/rebase_dyn
````

Strange, previous run was OK.

**Comment by jlupi** [comment]
Regarding the problem with open cygwin during update:
> The following DLLs couldn't be rebased because they were in use:

I may have a possible solution. We could add a babun upgrade option. This would simply do

````
PROXY= //take from .babunrc or from the user i.e. babun upgrade --proxy=host:port
open update.bat /proxy $PROXY && exit 
````
Open is a new alias to cygstart.
The update.bat would at the end call babun.bat to start the shell.

I just made a small PoC that does exactly this, looks like it is working.

I think this would be a better solution from the usability point of view - the user doesn't have to know where the upodate.bat is. The only problem is that the exit command would close the current window, there is no check if other instances are running.

@tombujok how do you see this? Do we want to include that in this release?

**Comment by tombujok** [comment]
Thanks for the update.
Every babun windows spawns a new /usr/bin/mintty process that you can see in any babun console by executing `ps` which will do the trick!

**Comment by jlupi** [comment]
all right so it will look more or less like this

````
open update.bat /proxy $PROXY && pkill mintty
````

**Comment by tombujok** [comment]
Fixed in 1.1.0. 
Release status: #266
